Add vLLM optimization guide and skill (#32)
intel/optimization-zone: software/vllm/README.md

> # vLLM on Intel Xeon Processors

Reviewer: Hi, Thanks for the contribution!

Consider adding a table of contents and breaking down the document into sections/subsections.

There are three broad topics being covered -
vLLM on Xeon guidance,
AI agent skill installation and benchmarking workflow.

Probably we can move the "Use with AI Coding Agents" towards the end, so we don't confuse and redirect someone who lands on the page for immediate tuning guidance.

Author: Added TOC

      optimization-zone-main/software/vllm/skill
```

### GitHub Copilot — User profile

```bash
mkdir -p ~/.copilot/skills/vllm-xeon-cpu
curl -L https://github.com/intel/optimization-zone/archive/refs/heads/main.tar.gz \
  | tar -xz --strip-components=4 -C ~/.copilot/skills/vllm-xeon-cpu \
      optimization-zone-main/software/vllm/skill
```

### Claude Code — Repo Workspace

```bash
mkdir -p .claude/skills/vllm-xeon-cpu
curl -L https://github.com/intel/optimization-zone/archive/refs/heads/main.tar.gz \
  | tar -xz --strip-components=4 -C .claude/skills/vllm-xeon-cpu \
      optimization-zone-main/software/vllm/skill
```

After install, invoke from chat with `/vllm-xeon-cpu` or let the agent auto-load the skill when your request matches keywords like "vLLM", "Xeon".

## Intel Xeon SLM/LLM Sizing Guidance

For guidance around SLM/LLM sizing on Intel Xeon CPUs, please see our Xeon Processor Advisor Tool & AI Software Catalog:

- [Cloud Intel Xeon AI Performance Advisor](https://xeonprocessoradvisor.intel.com/csp-ai-performance-advisor)
- [On-prem Intel Xeon AI Performance Advisor](https://xeonprocessoradvisor.intel.com/on-prem-ai-performance-advisor)

> - [Intel AI Software Catalog - Model Guidance](https://swcatalog.intel.com/models)

Reviewer: First two requires end user to provide email address.  Lets move the third bullet to the top, since that link has direct access to content.

Author: Done!

## vLLM Requirements Guidance

| Item | Guidance |
| --- | --- |
| OS | Linux |
| Python | 3.10 through 3.13 |
| vLLM | `v0.17.0` or newer |

> | Intel AMX Xeon CPU Flags | 4th Gen or newer with `amx_tile`, `amx_bf16`, and `amx_int8` for best BF16/INT8 performance |

Reviewer: ```suggestion
| Intel AMX related Xeon CPU Flags | 4th Gen Intel Xeon or newer with `amx_tile`, `amx_bf16`, and `amx_int8` for best BF16/INT8 performance |
```

Author: Fixed.

## Performance Guidance

| Setting | Guidance | Why it matters |
| --- | --- | --- |

> | `--dtype=bfloat16` | Use `bfloat16` on Intel Xeon with Intel AMX | Enables the preferred CPU dtype and AMX |

Reviewer: Does this enable the preferred CPU dtype and enables Intel AMX acceleration?

Author: Changed to "Selects the preferred vLLM CPU dtype for Intel AMX."

> | `VLLM_CPU_KVCACHE_SPACE` | `20` to `40` GiB or larger | Larger values allow more concurrency and context, but must fit per NUMA node. |

Reviewer 2: ```suggestion
| `VLLM_CPU_KVCACHE_SPACE` | `20` to `40` GiB or larger | Larger values allow more concurrency and context, but must fit into the memory capacity available per NUMA node. |
```

Author: Fixed

| `VLLM_CPU_OMP_THREADS_BIND` | `auto` | Binds OpenMP worker threads to NUMA-local cores. Use ranges such as `0-31\|32-63` for manual control, `auto` preferred. |

> | `VLLM_CPU_NUM_OF_RESERVED_CPU` | `1` | Sets a core for API serving, tokenization, networking, logging, and OS work. |

Reviewer: ```suggestion
| `VLLM_CPU_NUM_OF_RESERVED_CPU` | `1` | Reserves one core for API serving, tokenization, networking, logging, and OS work. |
```

Author: Fixed.

> | `--tensor-parallel-size` | Use default for single NUMA or set to NUMA node count | Keeps model shards close to local memory; current vLLM CPU releases do not support `6`. |

Reviewer: ```suggestion
| `--tensor-parallel-size` | Use the default for a single NUMA node, or set it to the NUMA node count | Keeps model shards close to local memory; current vLLM CPU releases do not support --tensor-parallel-size=6. |
```

Author: Fixed.

| `--max-num-batched-tokens` | Online: `2048`; offline: `4096` | Maximum number of batched tokens per iteration. Tune for prefill throughput and time to first token. |
| `--max-num-seqs` | Online: `128`; offline: `256` | Maximum number of sequences per iteration. Tune for decode throughput and inter-token latency. |
| `VLLM_CPU_SGL_KERNEL` | `0`, or try `1` for low-latency SLM serving | Experimental x86 small-batch kernels; requires AMX, BF16 weights, and compatible shapes. |

> 

Reviewer: Can you elaborate if the above performance guidance settings are environment settings? or some flags/arguments for a benchmark?

Author: Made is simpler to understand with additional explanation and two tables.

## Utility Tools

> Install the small host tools used by the commands below:

Reviewer 2: ```suggestion
Use the OS package manager to install the tools used by the commands below:
```

Author: Fixed.

```bash
sudo apt-get update
sudo apt-get install -y docker.io
sudo systemctl enable --now docker
sudo usermod -aG docker $USER
newgrp docker   # apply group without re-login
```

</details>

```bash
export HF_TOKEN=your_hf_token_here  # <<<=== Required for gated Hugging Face models and faster downloads.
export VLLM_VERSION=0.20.2          # <<<=== Update this for newer releases! Check! 
docker pull vllm/vllm-openai-cpu:v${VLLM_VERSION}-x86_64

docker run --rm \
  --name vllm-cpu \
  --security-opt seccomp=unconfined \
  --cap-add SYS_NICE \
  --shm-size=8g \
  -p 8000:8000 \
  -e HF_TOKEN="${HF_TOKEN}" \
  -e VLLM_CPU_KVCACHE_SPACE=20 \
  -e VLLM_CPU_OMP_THREADS_BIND=auto \
  -e VLLM_CPU_NUM_OF_RESERVED_CPU=1 \
  vllm/vllm-openai-cpu:v${VLLM_VERSION}-x86_64 \
  RedHatAI/Qwen3-4B-Instruct-2507-quantized.w8a8 \
  --dtype=bfloat16 \
  --max-num-batched-tokens 2048 \
  --max-num-seqs 128
```

`SYS_NICE` and `seccomp=unconfined` allow vLLM's NUMA memory policy calls inside Docker. Without them, serving can still work, but NUMA placement may be weaker and logs can show `get_mempolicy: Operation not permitted`.

## Validate the OpenAI-compatible endpoint

> **Open a new terminal or use a remote system.**

Reviewer 2: ```suggestion
**Open a new terminal and use the below command to test that the endpoint is available. Alternatively, you can connect from a remote system but make sure to substitute `localhost` for the server's address.**
```

Author: Fixed.

```bash
curl http://localhost:8000/v1/chat/completions \
  -H "Content-Type: application/json" \
  -d '{
    "model": "RedHatAI/Qwen3-4B-Instruct-2507-quantized.w8a8",
    "messages": [{"role": "user", "content": "Give three CPU inference tuning tips."}],
    "max_tokens": 128
  }'
```

## Benchmarking Guidance

This summarizes the official benchmarking and tuning guidance from the vLLM documentation, with a CPU focus. Always consult the [official benchmarking docs](https://docs.vllm.ai/en/latest/benchmarking/cli/) for the latest recommendations and tools.

> Start the Docker. If it is running in the foreground, open another terminal for these checks:

Reviewer: ```suggestion
Start the Docker container. If it is running in the foreground, open another terminal for these checks:
```

Author: Fixed

```bash

> # Docker path, because the server above is named vllm-cpu.

Reviewer: ```suggestion
# Docker path, because the container above is named vllm-cpu.
```

Author: Fixed.

docker exec vllm-cpu vllm collect-env

sudo curl -s http://localhost:8000/v1/models | jq .
SERVER_PID=$(pgrep -f 'vllm serve|api_server' | head -n 1)
numastat -p "${SERVER_PID}"
```

### Benchmark

> 

Reviewer: Can you add a few lines on what is vllm bench serve before the docker commands. Is it a subset of the Benchmark Suite you have there later?
Also, the execution context changes frequently between host/native install,  docker container, virtualenv. Someone may land up running host command in container or vice versa. 
For example, Concurrency sweep is provided for native environment. 
Also, are the utility tools, h/w validation etc. applicable before running the vLLM benchmark suite as well? or just for vllm bench serve?
Once you have sections/subsections hierarchy it will be more clear.

> Use `vllm bench serve` to measure TTFT, TPOT, and throughput against the running server. Warm up with `--num-warmups` to avoid measuring JIT compilation overhead.

Reviewer 2: Please provide the full form of TTFT and TPOT. I believe they're time to first token and time per output token right?

Author: Added

| Intel AMX related Xeon CPU Flags | 4th Gen Intel Xeon or newer with `amx_tile`, `amx_bf16`, and `amx_int8` for best BF16/INT8 performance |

### Utility Tools <!-- omit in toc -->

Use the OS package manager to install the tools used by the commands below:

```bash
sudo apt-get update
sudo apt-get install -y --no-install-recommends curl git jq numactl htop python3-venv python3-full g++ python3-dev
```

### Hardware Validation <!-- omit in toc -->

Validate the CPU model, core count, thread count, NUMA topology, and important flags such as `avx512f`, `avx2`, `amx_tile`, `amx_bf16`, `amx_int8`, and `avx512_bf16`.

```bash
lscpu | grep -E "Model name|Socket|Core|Thread|NUMA node|Flags"
lscpu | grep -E "avx512f|avx2|amx_(tile|bf16|int8)|avx512_bf16"
numactl --hardware
```

### Performance Guidance <!-- omit in toc -->

vLLM on CPU is tuned with two kinds of knobs, split into the two tables below.
**Environment variables** (`VLLM_CPU_*`) are set with a shell `export` or a Docker `-e` flag.
**Server CLI flags** (`--*`) are passed to `vllm serve` after the model name. Set each one in the
matching place shown in the Docker and benchmarking examples below. The **Example** column mirrors

> the runnable Docker command in [Validation Fast Path: Docker](#validation-fast-path-docker).

Reviewer: Validation Fast Path - Link doesn't work.

Author: Removed. I'm simplifying the README to keep it more concise.

**Open a new terminal and use the below command to test that the endpoint is available. Alternatively, you can connect from a remote system but make sure to substitute `localhost` for the server's address.**

```bash
curl http://localhost:8000/v1/chat/completions \
  -H "Content-Type: application/json" \
  -d '{
    "model": "RedHatAI/Qwen3-4B-Instruct-2507-quantized.w8a8",
    "messages": [{"role": "user", "content": "Give three CPU inference tuning tips."}],
    "max_tokens": 128
  }'
```

## Benchmarking Guidance

This summarizes the official benchmarking and tuning guidance from the vLLM documentation, with a CPU focus. Always consult the [official benchmarking docs](https://docs.vllm.ai/en/latest/benchmarking/cli/) for the latest recommendations and tools.

> **Mind the execution context.** The commands in this section run in one of three places: inside
> the **Docker container** (prefix with `docker exec vllm-cpu ...`), in a **native/host install**
> (`pip install vllm`, run directly), or inside a **Python virtualenv**. Run each command in the
> same environment where vLLM is installed — do not run a host command inside the container or a

> > container command on the host. The [Utility Tools](#utility-tools) and

Reviewer 2: Not sure how to fix this, but the link for Utility Tools is broken here. Looking at it, it should work and it actually works in the previewer in VS Code. I think the problem is that the comment next to the "Utility Tools" header is making issues. Can you remove all instances of "\<!-- omit in toc -->"?

Author: Removed. I'm simplifying the README to keep it more concise.

> > [Hardware Validation](#hardware-validation) steps are host-level checks: run them once on the

Reviewer 2: "Hardware Validation" link isn't working in the GitHub viewer either. I think this is also having issues with the "<\!-- omit in toc -->" comment in its header

Author: Removed. I'm simplifying README to keep it more concise.

```bash
for N in 10 50 100 200 500; do
  vllm bench serve \
    --model RedHatAI/Qwen3-4B-Instruct-2507-quantized.w8a8 \
    --dataset-name random \
    --random-input-len 128 \
    --random-output-len 128 \
    --num-prompts "${N}" \
    --num-warmups 5 \
    --request-rate inf \
    --save-result \
    --result-dir ./bench-results \
    --percentile-metrics ttft,tpot,itl
done
```

### Additional Testing & Tuning Methodology<!-- omit in toc -->

- Test with different input/output lengths to understand how the model performs under different prompt and generation sizes. For example, try `--random-input-len` and `--random-output-len` values of `64`, `128`, `256`, and `512`.
- Test with different user concurrency levels using `--num-prompts` values of `10`, `50`, `100`, `200`, and `500` with `--request-rate inf`.
- Use one known-good model and change one knob at a time. Track TTFT, TPOT, output tokens per second, requests per second, peak RSS, NUMA locality, and OOM events.

> - Vary only one of `VLLM_CPU_KVCACHE_SPACE`, `VLLM_CPU_OMP_THREADS_BIND`, `--max-num-batched-tokens`, `--max-num-seqs`, or `--block-size` per run. Compare results across runs using the saved JSON files in `./bench-results`.

Reviewer: Any recommendations for --block-size? Don't see it in the Server CLI flags

Author: We don't have specific Intel guidance on it. Removed. I'm simplifying the README to keep it more concise.
